### PR TITLE
Swift 2

### PR DESCRIPTION
**Pull request by cojoj**

### CODE REVIEW

File: XcodeServerSDK/Utils/HTTPUtils.swift [hunk 1]
```diff
@@ -30,15 +30,13 @@ public class HTTP {
             //try to cast into HTTP response
             if let httpResponse = response as? NSHTTPURLResponse {
                 
-                if error != nil {
-                    
+                guard error != nil else {
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: Are you sure you're using `guard` correctly? AFAIK you want to guard for something being true, in which case you want the control flow to continue. Otherwise, you want it to go into the scope.

In this case, the old code goes into the `//error in the networking stack` block iff `error != nil`. However, with your change with `guard`, now the block of code will be executed if `error == nil`, which is the opposite.

I think, I haven't used `guard` myself, but this is how I remember they work.

Comment (czechboy0) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: When it comes to changes, I think you want to change this line to `guard error == nil else {`, so that we guard against an error being present (in which case we want the `else` block to execute and return early.

File: XcodeServerSDK/Utils/HTTPUtils.swift [hunk 2]
```diff
@@ -30,15 +30,13 @@ public class HTTP {
...
                     //error in the networking stack
                     completion(response: httpResponse, body: nil, error: error)
                     return
                 }
                 
-                if data == nil {
-                    
+                guard data == nil else {
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: Here, on the other hand, we want to unwrap `data` and return if it's nil. So here I'd change it to `guard let data = data else {`. So that if `data` gets unwrapped successfully, the rest of the method will be able to use it unwrapped already.

File: XcodeServerSDK/Utils/HTTPUtils.swift [hunk 3]
```diff
@@ -53,13 +51,13 @@ public class HTTP {
                         
                     case let s where s.rangeOfString("application/json") != nil:
                         
-                        let (json: AnyObject!, error) = JSON.parse(data)
-                        let headers = httpResponse.allHeaderFields
+                        let (json, error) = JSON.parse(data!)
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: And the change above will let you get rid of the `!` here :)

File: XcodeServerSDK/Utils/HTTPUtils.swift [hunk 4]
```diff
@@ -53,13 +51,13 @@ public class HTTP {
...
+                        // let headers = httpResponse.allHeaderFields
                         completion(response: httpResponse, body: json, error: error)
                         
                     default:
                         //parse as UTF8 string
-                        let string = NSString(data: data, encoding: NSUTF8StringEncoding) as! String
+                        let string = NSString(data: data!, encoding: NSUTF8StringEncoding) as! String
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: And here as well.

File: XcodeServerSDK/Utils/HTTPUtils.swift [hunk 5]
```diff
@@ -92,7 +90,7 @@ public class HTTP {
             }
             
             
-        }).resume()
+        })?.resume()
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: Interesting, `dataTaskWithRequest` changed to returning an optional? In that case, I think it's a good place for another guard (we want to still call the completion block if `nil` is returned, otherwise it would never be called).

So higher up, try `guard let task = self.session.dataTaskWithRequest(...) else {` and call completion block with some error in the else block. Then, underneath all that, start the task with `task.resume()`. Does that make sense?

File: XcodeServerSDK/Utils/JSON.swift [hunk 6]
```diff
@@ -15,34 +15,32 @@ public protocol JSONSerializable {
...
+        return (object as? [AnyObject], error)
     }
     
-    public class func parse(url: NSURL) -> (AnyObject!, NSError!) {
+    public class func parse(url: NSURL) -> (AnyObject?, NSError?) {
         
-        var error: NSError?
-        if let data = NSData(contentsOfURL: url, options: NSDataReadingOptions.allZeros, error: &error) {
+        do {
+            let data = try NSData(contentsOfURL: url, options: NSDataReadingOptions(rawValue: 0))
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/JSON.swift:31]: :+1: Nice.

Just btw, can you put the options back to `NSDataReadingOptions.allZeros`?

File: XcodeServerSDK/Utils/JSON.swift [hunk 7]
```diff
@@ -15,34 +15,32 @@ public protocol JSONSerializable {
...
-    public class func parse(data: NSData) -> (AnyObject!, NSError!) {
-        
-        var parsingError: NSError?
-        if let obj: AnyObject = NSJSONSerialization.JSONObjectWithData(data, options: NSJSONReadingOptions.AllowFragments, error: &parsingError) {
-            return (obj as AnyObject!, nil)
-        } else {
-            return (nil, parsingError)
+    public class func parse(data: NSData) -> (AnyObject?, NSError?) {
+        do {
+            let obj = try NSJSONSerialization.JSONObjectWithData(data, options: .AllowFragments)
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/JSON.swift:40]: Nice, this does look much better.

File: XcodeServerSDK/Utils/JSON.swift (hunk 6 above)
Comment (cojoj) [inline XcodeServerSDK/Utils/JSON.swift:31]: It was complaining that there is nothing like `allZeros`

File: XcodeServerSDK/Utils/Logging.swift [hunk 8]
```diff
@@ -50,7 +50,7 @@ public class ConsoleLogger: Logger {
     }
     
     public func log(message: String) {
-        println(message)
+        print(message)
```
Comment (czechboy0) [inline XcodeServerSDK/Utils/Logging.swift:53]: Why the change from `println` to `print`? I think we still want that newline...

Comment (cojoj) [inline XcodeServerSDK/Utils/Logging.swift:53]: There no more `println()` it was changed to `print()` but if we want a new line it's `print(value, appendNewline)`

File: XcodeServerSDK/Utils/JSON.swift (hunk 6 above)
Comment (czechboy0) [inline XcodeServerSDK/Utils/JSON.swift:31]: Oh ok, must have been a Swift 2 change. :+1:

File: XcodeServerSDK/Utils/Logging.swift (hunk 8 above)
Comment (czechboy0) [inline XcodeServerSDK/Utils/Logging.swift:53]: You have to be joking... :laughing: Way to make Swift less usable, wow. Ok, keep it this way then, I'll take a look once we merge it.

**Comment by czechboy0** [comment]
Can you also please re-apply the changes in https://github.com/czechboy0/XcodeServerSDK/commit/41bfb8a3f3b0df6a58aa0a2fc722c6091207781e? Because you moved the `XcodeServerConfig` into a separate file, so the merge won't apply that cleanup change anymore (you'll have to do it manually).

File: XcodeServerSDK/Utils/HTTPUtils.swift (hunk 3 above)
Comment (cojoj) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: That what I thought but...
zrzut ekranu 2015-06-16 o 21 51 11

Comment (cojoj) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: That's because we're only checking against `nil` but not doing optional unwrapping... Will change it!

File: XcodeServerSDK/Utils/Logging.swift (hunk 8 above)
Comment (czechboy0) [inline XcodeServerSDK/Utils/Logging.swift:53]: Turns out `true` for `appendNewline` is actually the default. So all good in the hood.

Comment (cojoj) [inline XcodeServerSDK/Utils/Logging.swift:53]: 👍

File: XcodeServerSDK/Utils/HTTPUtils.swift (hunk 5 above)
Comment (cojoj) [inline XcodeServerSDK/Utils/HTTPUtils.swift]: It makes a perfect sense!

**Comment by cojoj** [comment]
@czechboy0 I've changed/repaired thing you've pointed so... Your job now to review it one more time. I'm still pretty skeptical because we don't have a proper test coverage to check of everything is working properly.

**Comment by czechboy0** [comment]
@cojoj The lack of tests is definitely something to fix sooner rather than later. But up until now the only thing I could have done was to not release the project until I have time to write them.

I would definitely appreciate if you started adding some tests in the suite.

File: XcodeServerSDK/XcodeServerConfig.swift [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        self.availabilityState = .Unchecked
+    }
+    
+    public required init?(json: NSDictionary) {
+        
+        self.availabilityState = .Unchecked
+        
+        if
+            let host = json.optionalStringForKey("host"),
+            let apiVersionString = json.optionalStringForKey("api_version")
```
Comment (czechboy0) [inline XcodeServerSDK/XcodeServerConfig.swift]: This should go as well.

**Comment by czechboy0** [comment]
I think you missed the issues in `HTTPUtils.swift`, I think there are some logic problems with `guard` (see inline comments).

File: XcodeServerSDK/XcodeServerConfig.swift (hunk 9 above)
Comment (cojoj) [inline XcodeServerSDK/XcodeServerConfig.swift]: ??

**Comment by cojoj** [comment]
@czechboy0 `completion` you're talking about should return only `NSError`? Of what type?

File: XcodeServerSDK/XcodeServerConfig.swift (hunk 9 above)
Comment (czechboy0) [inline XcodeServerSDK/XcodeServerConfig.swift]: This is not needed anymore :)

Comment (czechboy0) [inline XcodeServerSDK/XcodeServerConfig.swift]: This: `let apiVersionString = json.optionalStringForKey("api_version")`

Comment (cojoj) [inline XcodeServerSDK/XcodeServerConfig.swift]: This is done... Sorry I missed this one 😑

**Comment by czechboy0** [comment]
@cojoj What I mean that in the `else` block, you should call something like `completion(response: nil, body:nil, error:Error.withInfo("Could't create request"))` and then `return`.

**Comment by cojoj** [comment]
@czechboy0 IMHO we should drop returning `NSError`s and switch to `throw`ing them but it's probably a bigger portion of refactoring.

**Comment by czechboy0** [comment]
Cool, now just please fix the logic issues in `HTTPUtils.swift` and we'll be able to merge :)

**Comment by cojoj** [comment]
I've changed them and reverted them back... Deserve some kind of award 😓

**Comment by czechboy0** [comment]
Awesome!

**Comment by czechboy0** [comment]
You get a :fireworks:

**Comment by czechboy0** [comment]
:clap: Thanks!

**Comment by cojoj** [comment]
Weeeeeeee!!! 🎆

Giphy
